File: 2020/08/puzzle.py
with open('/Users/mailund/Projects/adventofcode/2020/08/input.txt') as f:
    prog = [tuple(cast(x) for cast,x in zip((str,int), inst.split()))
            for inst in f.readlines()]

class State(object):
    def __init__(self):
        self.instp = 0 # instruction pointer
        self.acc_ = 0  # accumulator
    def nop(self, operand):
        self.instp += 1
    def acc(self, operand):
        self.acc_ += operand
        self.instp += 1
    def jmp(self, operand):
        self.instp += operand

def run_prog(prog):
    visited = set()
    state = State()
    ops = {op: getattr(state,op) for op in dir(State) if not op.startswith('__')}
    while state.instp < len(prog):
        if state.instp in visited: return state.acc_
        visited.add(state.instp)
        op, operand = prog[state.instp]
        ops[op](operand)

print(f"Puzzle #1: {run_prog(prog)}")


def run_prog(prog):
    visited = set()
    state = State()
    ops = {op: getattr(state,op) for op in dir(State) if not op.startswith('__')}
    while state.instp < len(prog):
        if state.instp in visited: return ('Abort', state.acc_)
        visited.add(state.instp)
        op, operand = prog[state.instp]
        ops[op](operand)
    return ('Success', state.acc_)

def alternative_progs(prog):
    for i, (op, operand) in enumerate(prog):
        if op == 'nop':
            prog[i] = ('jmp', operand)
            yield prog
            prog[i] = ('nop', operand)
        if op == 'jmp':
            prog[i] = ('nop', operand)
            yield prog
            prog[i] = ('jmp', operand)

for alt_prog in alternative_progs(prog[:]):
    res, acc = run_prog(alt_prog)
    if res == 'Success':
        print(f"Puzzle #2: {acc}")
        break

# hack for non-local return
class CallCC(Exception):
    def __init__(self, resval):
        self.resval = resval

def callcc(f, *args):
    try:
        f(*args)
    except CallCC as x:
         return x.resval

def locate_index(prog, seen, i, j):
    if i == len(prog): raise CallCC(j)
    if i in seen: return

    seen.add(i)
    op, operand = prog[i]
    if op == 'acc':
        locate_index(prog, seen, i+1, j)
    elif op == 'nop':
        locate_index(prog, seen, i + 1, j)
        # switch to jmp
        if j is None:
            locate_index(prog, seen, i + operand, i)
    else: # op is jmp
        locate_index(prog, seen, i + operand, j)
        # switch to nop
        if j is None:
            locate_index(prog, seen, i + 1, i)
    seen.remove(i)

def change_index(prog):
    idx = callcc(locate_index, prog, set(), 0, None)
    op, operand = prog[idx]
    if op == 'nop': prog[idx] = 'jmp', operand
    else:           prog[idx] = 'nop', operand

change_index(prog)
print(f"Puzzle #2: {run_prog(prog)}")

File: 2020/08/puzzle.c
#include <stdio.h>
#include <string.h>

struct command { int inc_instp; int inc_acc; };
struct state   { int instp;     int acc;     };
#define CMD(ii, ia) \
  (struct command){ .inc_instp = (ii), .inc_acc = (ia) }

#define N 1000 // the program is shorter than this...
struct command PROG[N];
char visited[N];

static int parse_asm(void)
{
    int n = 0; char buf[4]; int operand;
    while (scanf("%s %d\n", buf, &operand) == 2) {
        if (strcmp(buf, "nop") == 0)      PROG[n++] = CMD(1, 0);
        else if (strcmp(buf, "acc") == 0) PROG[n++] = CMD(1, operand);
        else /* jmp */                    PROG[n++] = CMD(operand, 0);
    }
    return n;
}

struct res { enum { SUCCESS, FAILURE } status; int res; };
#define SUCC(state) (struct res){ .status = SUCCESS, .res = state.acc }
#define FAIL(state) (struct res){ .status = FAILURE, .res = state.acc }

#define UPD(s, cmd) /* important to upd acc before instp */ \
  do { s.acc += cmd.inc_acc; s.instp += cmd.inc_instp; } while(0)

static struct res run(int n)
{
    struct state state = { .instp = 0, .acc = 0 };
    while (state.instp != n) {
        if (visited[state.instp]) return FAIL(state);
        visited[state.instp] = 1;
        UPD(state, PROG[state.instp]);
    }
    return SUCC(state);
}

int main(void)
{
    struct res res = run(parse_asm());
    switch (res.status) {
        case SUCCESS: printf("SUCCESS: %d\n", res.res); break;
        case FAILURE: printf("FAILURE: %d\n", res.res); break;
    }
    return 0;
}
